### PR TITLE
tikv: log slow coprocessor task in detail

### PR DESCRIPTION
**Pull request by coocood**
Add transaction startTS, storeAddr, ExecDetail to the log to make debug easier.

### CODE REVIEW

**Comment by coocood** [comment]
@zz-jason @shenli PTAL

File: store/tikv/coprocessor.go [hunk 1]
```diff
@@ -628,6 +628,57 @@ func (worker *copIteratorWorker) handleTaskOnce(bo *Backoffer, task *copTask, ch
...
+const (
+	minLogBackoffTime       = 100
+	minLogKVProcessTime     = 100
+	minLogKVWaitTime        = 200
+	minLogScanProcessedInfo = 100000
+	minLogScanTotalInfo     = 200000
+)
+
+func (worker *copIteratorWorker) logTimeCopTask(costTime time.Duration, task *copTask, bo *Backoffer, resp *tikvrpc.Response) {
+	logStr := fmt.Sprintf("[TIME_COP_TASK] cop_resp_time:%s, cop_start_ts:%d, cop_region_id:%d, cop_store_addr:%s", costTime, worker.req.StartTs, task.region.id, task.storeAddr)
```
Comment (shenli) [inline store/tikv/coprocessor.go]: Add time unit for the cop_resp_time.

**Comment by shenli** [comment]
@AndreMouche PTAL

File: store/tikv/coprocessor.go (hunk 1 above)
Comment (coocood) [reply store/tikv/coprocessor.go]: It's `time.Duration` type, which unit is included.

Comment (zz-jason) [reply store/tikv/coprocessor.go]: s/cop_start_ts/txnID/ ?

Comment (zz-jason) [reply store/tikv/coprocessor.go]: I think we don't need to print the 'cop_' prefix

File: store/tikv/coprocessor.go [hunk 2]
```diff
@@ -628,6 +628,57 @@ func (worker *copIteratorWorker) handleTaskOnce(bo *Backoffer, task *copTask, ch
...
+	if task.cmdType == tikvrpc.CmdCopStream {
+		detail = resp.CopStream.ExecDetails
+	} else {
+		detail = resp.Cop.ExecDetails
+	}
+	if detail != nil {
+		if detail.HandleTime != nil {
+			processMs := detail.HandleTime.ProcessMs
+			waitMs := detail.HandleTime.WaitMs
+			if processMs > minLogKVProcessTime {
```
Comment (zz-jason) [inline store/tikv/coprocessor.go]: @AndreMouche Does coprocessor also has a threshold to fill `HandleTime.ProcessMs` and `HandleTime.WaitMs` for slow cop requests ?

**Comment by coocood** [comment]
@zz-jason PTAL

File: store/tikv/coprocessor.go (hunk 2 above)
Comment (AndreMouche) [reply store/tikv/coprocessor.go]: YES

**Comment by coocood** [comment]
@zz-jason @AndreMouche PTAL

File: store/tikv/coprocessor.go (hunk 2 above)
Comment (zz-jason) [reply store/tikv/coprocessor.go]: @coocood Maybe we just need to print `HandleTime.ProcessMs` and `HandleTime.WaitMs` once they are set, no need to use another thresholds like `minLogKVProcessTime` and `minLogKVWaitTime`

Comment (coocood) [reply store/tikv/coprocessor.go]: After we set the `HandleTime` and `ScanDetial` to `true`, every request will return them.

File: store/tikv/coprocessor.go [hunk 3]
```diff
@@ -628,6 +631,58 @@ func (worker *copIteratorWorker) handleTaskOnce(bo *Backoffer, task *copTask, ch
...
+		backoffTypes := strings.Replace(fmt.Sprintf("%v", bo.types), " ", ",", -1)
+		logStr += fmt.Sprintf(" backoff_ms:%d backoff_types:%s", bo.totalSleep, backoffTypes)
+	}
+	var detail *kvrpcpb.ExecDetails
+	if task.cmdType == tikvrpc.CmdCopStream {
+		detail = resp.CopStream.ExecDetails
+	} else {
+		detail = resp.Cop.ExecDetails
+	}
+	if detail != nil {
```
Comment (AndreMouche) [inline store/tikv/coprocessor.go]: There are so many `if` in `if`..  Could we `log.Info` through `defer` function, and return directly `if detail==nil`?

**Comment by coocood** [comment]
@shenli PTAL

File: store/tikv/coprocessor.go (hunk 2 above)
Comment (tiancaiamao) [reply store/tikv/coprocessor.go]: When will we set `ScanDetail` to `true` ? @coocood

Comment (coocood) [reply store/tikv/coprocessor.go]: ScanDetail is not set to true now. it's set only when the duration exceeds the threshold(1s) in TiKV.

File: store/tikv/coprocessor.go (hunk 3 above)
Comment (coocood) [reply store/tikv/coprocessor.go]: I think defer is used for release resource.
I looks weird if the normal execution code is done in defer.

Comment (zhexuany) [reply store/tikv/coprocessor.go]: Agree with Yuxing.

File: store/tikv/coprocessor.go [hunk 4]
```diff
@@ -604,21 +605,22 @@ func (worker *copIteratorWorker) handleTaskOnce(bo *Backoffer, task *copTask, ch
 			IsolationLevel: pbIsolationLevel(worker.req.IsolationLevel),
 			Priority:       kvPriorityToCommandPri(worker.req.Priority),
 			NotFillCache:   worker.req.NotFillCache,
+			HandleTime:     true,
```
Comment (zhexuany) [inline store/tikv/coprocessor.go]: I feel like "EnableHandleTime" is better.

Comment (coocood) [reply store/tikv/coprocessor.go]: This is defined in protobuf, hard to change.

File: store/tikv/coprocessor.go (hunk 2 above)
Comment (zz-jason) [reply store/tikv/coprocessor.go]: I still think we should remove these `minXXXX` thresholds defined in this file and do not check them.

File: store/tikv/coprocessor.go [hunk 5]
```diff
@@ -628,6 +630,58 @@ func (worker *copIteratorWorker) handleTaskOnce(bo *Backoffer, task *copTask, ch
...
+			logStr = appendScanDetail(logStr, "data", detail.ScanDetail.Data)
+			logStr = appendScanDetail(logStr, "lock", detail.ScanDetail.Lock)
+		}
+	}
+	log.Info(logStr)
+}
+
+func appendScanDetail(logStr string, columnFamily string, scanInfo *kvrpcpb.ScanInfo) string {
+	if scanInfo != nil {
+		if scanInfo.Total > minLogScanTotalInfo {
```
Comment (shenli) [inline store/tikv/coprocessor.go]: Why we need to check this?

File: store/tikv/coprocessor.go (hunk 2 above)
Comment (coocood) [reply store/tikv/coprocessor.go]: TiDB should control what to log, without depending on TiKV's implementation.

File: store/tikv/coprocessor.go (hunk 5 above)
Comment (coocood) [reply store/tikv/coprocessor.go]: I think we should not log messages that is not interesting.

File: store/tikv/coprocessor.go (hunk 2 above)
Comment (coocood) [reply store/tikv/coprocessor.go]: @zz-jason If we log everything, the most important ones will be obscure.

**Comment by shenli** [comment]
LGTM

**Pull request merged**